crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2384 @@
+ * @return index in AEAD table or 0 if error
+ */
+static uint16_t aead_iana2index(uint16_t codepoint)
+{
+    uint16_t naeads = OSSL_NELEM(hpke_aead_tab);
+    uint16_t i = 0;
+
+    for (i = 0; i != naeads; i++) {
+        if (hpke_aead_tab[i].aead_id == codepoint) {
+            return i;
+        }
+    }
+    return 0;
+}
+
+/*
+ * @brief map from IANA codepoint to KEM table index
+ *
+ * @param codepoint should be an IANA code point
+ * @return index in KEM table or 0 if error
+ */
+static uint16_t kem_iana2index(uint16_t codepoint)
+{
+    uint16_t nkems = OSSL_NELEM(hpke_kem_tab);
+    uint16_t i = 0;
+
+    for (i = 0; i != nkems; i++) {
+        if (hpke_kem_tab[i].kem_id == codepoint) {
+            return i;
+        }
+    }
+    return 0;
+}
+
+/*
+ * @brief map from IANA codepoint to AEAD table index
+ *
+ * @param codepoint should be an IANA code point
+ * @return index in AEAD table or 0 if error
+ */
+static uint16_t kdf_iana2index(uint16_t codepoint)
+{
+    uint16_t nkdfs = OSSL_NELEM(hpke_kdf_tab);
+    uint16_t i = 0;
+
+    for (i = 0; i != nkdfs; i++) {
+        if (hpke_kdf_tab[i].kdf_id == codepoint) {
+            return i;
+        }
+    }
+    return 0;
+}
+
+/*
+ * @brief Check if kem_id is ok/known to us
+ * @param kem_id is the externally supplied kem_id
+ * @return 1 for good, not 1 for error
+ */
+static int hpke_kem_id_check(uint16_t kem_id)
+{
+    switch (kem_id) {
+    case OSSL_HPKE_KEM_ID_P256:
+    case OSSL_HPKE_KEM_ID_P384:
+    case OSSL_HPKE_KEM_ID_P521:
+    case OSSL_HPKE_KEM_ID_X25519:
+    case OSSL_HPKE_KEM_ID_X448:
+        break;
+    default:
+        return 0;
+    }
+    return 1;
+}
+
+/*
+ * @brief check if KEM uses NIST curve or not
+ * @param kem_id is the externally supplied kem_id
+ * @return 1 for NIST, 0 for good-but-non-NIST, other otherwise
+ */
+static int hpke_kem_id_nist_curve(uint16_t kem_id)
+{
+    switch (kem_id) {
+    case OSSL_HPKE_KEM_ID_P256:
+    case OSSL_HPKE_KEM_ID_P384:
+    case OSSL_HPKE_KEM_ID_P521:
+        return 1;
+    default:
+        return 0;
+    }
+    return 0;
+}
+
+/*
+ * @brief hpke wrapper to import NIST curve public key as easily as x25519/x448
+ *
+ * @param libctx is the context to use
+ * @param propq is a properties string
+ * @param curve is the curve NID
+ * @param gname is the curve groupname
+ * @param buf is the binary buffer with the (uncompressed) public value
+ * @param buflen is the length of the private key buffer
+ * @return a working EVP_PKEY * or NULL
+ */
+static EVP_PKEY * EVP_PKEY_new_raw_nist_public_key(OSSL_LIB_CTX *libctx,
+                                                   const char *propq,
+                                                   int curve,
+                                                   const char *gname,
+                                                   const unsigned char *buf,
+                                                   size_t buflen)
+{
+    int erv = 1;
```

Comment:
If you set this to 0 instead, Then you dont need to propagate the erv = 0 everywhere..

Reply:
no prob with that change